subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformWithDependenciesIntegrationTest.groovy

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.resolve.transform
+
+import org.gradle.integtests.fixtures.AbstractHttpDependencyResolutionTest
+import org.gradle.util.ToBeImplemented
+
+class ArtifactTransformWithDependenciesIntegrationTest extends AbstractHttpDependencyResolutionTest {
+
+    def setup() {
+        // TODO LJA Investigate, this makes little sense
+        executer.expectDeprecationWarning()
+
+        settingsFile << """
+            rootProject.name = 'transform-deps'
+            include 'lib'
+            include 'app'
+        """
+
+        file('lib/src/main/java/test/MyClass.java') << """
+package test;
+
+public class MyClass {
+    public static void main(String[] args) {
+        System.out.println("Hello world!");
+    }
+}
+"""
+
+        buildFile << """
+def artifactType = Attribute.of('artifactType', String)
+
+allprojects {
+    repositories {
+        mavenCentral()
+    }
+}
+project(':lib') {
+    apply plugin: 'java-library'
+
+    dependencies {
+        api 'org.slf4j:slf4j-api:1.7.25'
+    }
+}
+
+project(':app') {
+    apply plugin: 'java'
+    
+    dependencies {
+        implementation 'junit:junit:4.11'
+        implementation project(':lib')
+    }
+
+    dependencies {
+        registerTransform {
+            from.attribute(artifactType, 'jar')
+            to.attribute(artifactType, 'size')
+            artifactTransform(FileSizer)
+        }
+    }
+}
+
+import javax.inject.Inject
+import org.gradle.api.artifacts.transform.ArtifactTransformDependencies
+
+class FileSizer extends ArtifactTransform {
+
+    ArtifactTransformDependencies artifactDependencies
+
+    @Inject
```

Comment:
I feels really weird to me that we inject the artifact dependencies, but that the actual artifact being transformed is passed as an argument to the `transform` method. I know this is done because not all transforms need it, but just wanted to highlight the pattern is a bit strange.

Reply:
Do you have suggestions about how to do this differently? What other options do we have?

We need a declarative way that can be inspected by Gradle without actually executing the transform to decide whether or not this information is going to be used as input. That means we can't make it always available.

Another option is what we do with `IncrementalTaskInputs`, i.e. we inject it into the `@TaskAction` method. I'm not a fan of that approach either, because it leads to weird questions like "what order do you execute things when they have multiple `@TaskAction` methods?", or "what if there's no `@TaskAction` method?".

The only downside I see with injected services is that there is a discoverability problem, but that problem exists with the `@TaskAction`-injected approach as well, and can be solved in several ways (e.g. by added documentation, implement a common `UserInjectableService` interface etc.).

Basically, I see three options to expose this information to a transform:

* expose always via some baked in parameter of method
* inject via constructor parameter
* inject via `@TransformAction` method parameter

Do you know of other solutions we could use? Of the above I think constructor injection is the only one worth considering.